File: lib/languages/zig/function-arguments.zig
// Basic function arguments
fn add(a: i32, b: i32) i32 {
    return a + b;
}

// Passing by reference (pointer)
fn increment(value: *i32) void {
    value.* += 1;
}

// Slices for variable-length arrays
fn sum(numbers: []const i32) i32 {
    var total: i32 = 0;
    for (numbers) |n| {
        total += n;
    }
    return total;
}

// Optional arguments
fn greet(name: []const u8, greeting: ?[]const u8) []const u8 {
    const default_greeting = "Hello";
    const actual_greeting = greeting orelse default_greeting;
    return actual_greeting ++ ", " ++ name ++ "!";
}

// Calling with optional arguments
const message1 = greet("Alice", null);  // "Hello, Alice!"
const message2 = greet("Bob", "Hi");    // "Hi, Bob!"

// Struct for named arguments
const Config = struct {
    name: []const u8,
    age: ?u32 = null,
    email: ?[]const u8 = null,
};

fn createUser(config: Config) User {
    return User{
        .id = generateId(),
        .name = config.name,
        .age = config.age orelse 0,
        .email = config.email orelse "",
    };
}

// Calling with named arguments
const user = createUser(.{
    .name = "Alice",
    .age = 30,
    .email = "alice@example.com",
});

// Variadic arguments with comptime
fn print(comptime fmt: []const u8, args: anytype) void {
    const stdout = std.io.getStdOut().writer();
    stdout.print(fmt, args) catch unreachable;
}

// Calling with variadic arguments
print("Name: {s}, Age: {d}\n", .{ "Alice", 30 });

// Helper types for the examples
const User = struct {
    id: []const u8,
    name: []const u8,
    age: u32,
    email: []const u8,
};

fn generateId() []const u8 {
    return "user123";
}

File: lib/languages/rust/function-arguments.rs
// Basic function arguments
fn add(a: i32, b: i32) -> i32 {
   a + b
}

// Passing by reference
fn increment(x: &mut i32) {
   *x += 1;
}

// Slices for variable-length arrays
fn sum(numbers: &[i32]) -> i32 {
   numbers.iter().sum()
}

// Default function arguments (via Option)
fn greet(name: &str, greeting: Option<&str>) -> String {
   let greeting = greeting.unwrap_or("Hello");
   format!("{}, {}!", greeting, name)
}

// Calling with default arguments
let message = greet("Alice", None);  // "Hello, Alice!"
let custom = greet("Bob", Some("Hi"));  // "Hi, Bob!"

// Named arguments (via structs)
struct Config {
   name: String,
   age: Option<u32>,
   email: Option<String>,
}

fn create_user(config: Config) -> User {
   User {
       id: generate_id(),
       name: config.name,
       age: config.age.unwrap_or(0),
       email: config.email.unwrap_or_else(|| String::from("")),
   }
}

// Calling with named arguments
let user = create_user(Config {
   name: String::from("Alice"),
   age: Some(30),
   email: Some(String::from("alice@example.com")),
});

// Variadic arguments (via macros)
macro_rules! sum_all {
   ($($x:expr),*) => {
       {
           let mut total = 0;
           $(total += $x;)*
           total
       }
   };
}

let total = sum_all!(1, 2, 3, 4);  // 10

// Helper types for the examples
struct User {
   id: String,
   name: String,
   age: u32,
   email: String,
}

fn generate_id() -> String {
   String::from("user123")
}

